hw-2

crafty-ezhik opened the pull request.

crafty-ezhik committed:
feat: Обновлен taslfile

Добавлены следующие задачи:
1. Установка mockery
2. Генерация моков
3. Запуск тестов
4. Подсчет покрытия кода
5. Необходимые переменные для работы

crafty-ezhik committed:
feat: Сделаны модели, интерфейс, pass-методы для репозитория

Commit:
feat: Имлементирован интерфейс InventoryService

Commit:
feat: Сделаны конверторы и модель service слоя

crafty-ezhik committed:
feat: Имплементирован InventoryService, добавлены модели и конверторы

crafty-ezhik committed:
feat: Добавлен ctx в слой service и repository

crafty-ezhik committed:
feat: Имплементирован InventoryAPI, добавлен конвертор

crafty-ezhik committed:
style: сортировка импортов

crafty-ezhik committed:
feat: Имлементирован слой service

crafty-ezhik committed:
feat: Вынос фильтрации в слой repository

crafty-ezhik committed:
fix: Изменено имя пакета в сервисном слое

crafty-ezhik committed:
fix: Добавлена генерация деталей при старте сервера

Commit:
feat: Payament Service разбит на слои

Commit:
feat: Добавлены и имплементированы gRPC клиенты

Commit:
style: форматирование

Commit:
feat: Добавлены ручки в api слое

Commit:
feat: Реализованы конверторы для gRPC клиентов

Commit:
fix: Изменено имя пакета у файлов сервиса

Commit:
feat: Добавлены репо и сервис интерфейсы

Commit:
feat: Добавлено описание моделей

Commit:
fix: Сделаны методы интерфейсов сервиса и репозитория

crafty-ezhik committed:
feat: Имплементирован OrderService

crafty-ezhik committed:
feat: Добавлены конверторы и модели в слой repository

crafty-ezhik committed:
feat: добавлены enums и конверторы в service слой

crafty-ezhik committed:
fix: Изменен тип парметра paymentMethod у ListParts

crafty-ezhik committed:
feat: Имплементирован OrderRepository

crafty-ezhik committed:
feat: Добавлена 408 ошибка в openAPI декларацию

crafty-ezhik committed:
feat: Добавлен 408 код ко всем ручка в openAPI

crafty-ezhik committed:
fix: Исправления

1. Убран параметр userID у метода Create в слое service
2. Добавлен возврат totalPrice
3. Исправлено имя ошибки

crafty-ezhik committed:
feat: Имплементирован api слой

crafty-ezhik committed:
refactor: отредактирован main.go

crafty-ezhik committed:
feat: Сгенерированы моки для всех сервисов

crafty-ezhik committed:
fix: Убран enums из слоя репозитория.

Теперь используется значенич из сервисного слоя

crafty-ezhik committed:
chore: обновление зависимостей

crafty-ezhik committed:
test: Сделаны тесты для PaymentService

crafty-ezhik committed:
test: Написаны тесты для слоя service [InventoryService]

crafty-ezhik committed:
fix:  Перенесено формирование orderUUID на сторону репозитория

crafty-ezhik committed:
test: Написаны тесты для OrderService слой service

crafty-ezhik committed:
style: Сортировка ипортов и форматирование

crafty-ezhik committed:
fix: добавлена проверка на nil в конверторах

crafty-ezhik committed:
fix: Убрана константа в тесте

crafty-ezhik committed:
test: Написаны тесты для api слоя InventoryService

crafty-ezhik committed:
test: Написаны тесты для api слоя OrderService

crafty-ezhik committed:
feat: Добавлена обработка ошибки, если заказ оплачен или отменен

crafty-ezhik committed:
docs: Добавлен бейдж покрытия тестами

crafty-ezhik committed:
docs: обновлен Readme.md

crafty-ezhik committed:
Обновление покрытия тестами только в main

mistandok commented on inventory/internal/api/inventory/v1/get.go:
```diff
@@ -0,0 +1,39 @@
+
+	part, err := a.inventoryService.Get(ctx, partID)
+	if err != nil {
+		if errors.Is(err, model.ErrPartNotFound) {
```
suggestion(non-blocking): можешь делать, а можешь не делать, это просто рекомендация, как можно поработать с ошибками и облегчить их преобразование к grpc (пример нейронкой сгенерировал, могут быть косяки, нужно проверять дополнительно). Ты сейчас ошибки кастуешь в каждом handler, а можно для этого бахнуть интерцептор. Но тогда надо немного доработать сами ошибки

```go
// shared/pkg/errors/business_error.go
package errors

import (
	"errors"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// ErrorCode represents business error codes
type ErrorCode int64

const (
	NotFoundErrCode ErrorCode = iota
	InvalidArgumentErrCode
)

// businessError represents a structured business error
type businessError struct {
	code ErrorCode
	err  error
}

func (b *businessError) Error() string {
	if b.err != nil {
		return b.err.Error()
	}
	return "unknown business error"
}

func (b *businessError) Unwrap() error {
	return b.err
}

func (b *businessError) Code() ErrorCode {
	return b.code
}

// Constructor functions for different error types
func NewNotFoundError(err error) *businessError {
	return &businessError{
		code: NotFoundErrCode,
		err:  err,
	}
}

func NewInvalidArgumentError(err error) *businessError {
	return &businessError{
		code: InvalidArgumentErrCode,
		err:  err,
	}
}

// GetBusinessError returns businessError if err is a business error, nil otherwise
func GetBusinessError(err error) *businessError {
	var businessErr *businessError
	if errors.As(err, &businessErr) {
		return businessErr
	}
	return nil
}

// errorCodeToGRPCCode maps business error codes to gRPC codes
func errorCodeToGRPCCode(code ErrorCode) codes.Code {
	switch code {
	case NotFoundErrCode:
		return codes.NotFound
	case InvalidArgumentErrCode:
		return codes.InvalidArgument
	default:
		return codes.Unknown
	}
}

// BusinessErrorToGRPCStatus converts businessError to gRPC status
func BusinessErrorToGRPCStatus(err *businessError) *status.Status {
	grpcCode := errorCodeToGRPCCode(err.Code())
	return status.New(grpcCode, err.Error())
}

// shared/pkg/interceptors/error_interceptor.go
package interceptors

import (
	"context"
	"log"

	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"

	"github.com/green-vlad/mcb/shared/pkg/errors"
)

// UnaryErrorInterceptor handles error conversion for unary RPC calls
func UnaryErrorInterceptor() grpc.UnaryServerInterceptor {
	return func(
		ctx context.Context,
		req interface{},
		info *grpc.UnaryServerInfo,
		handler grpc.UnaryHandler,
	) (interface{}, error) {
		resp, err := handler(ctx, req)
		if err != nil {
			return resp, convertError(err, info.FullMethod)
		}
		return resp, nil
	}
}

// convertError converts business errors to appropriate gRPC errors
func convertError(err error, method string) error {
	// Check if it's a businessError
	if businessErr := errors.GetBusinessError(err); businessErr != nil {
		grpcStatus := errors.BusinessErrorToGRPCStatus(businessErr)
		log.Printf("BusinessError in method %s: code=%d, message=%s", 
			method, businessErr.Code(), businessErr.Error())
		return grpcStatus.Err()
	}

	// Check if it's already a gRPC status error
	if _, ok := status.FromError(err); ok {
		return err
	}

	// For unknown errors, return internal error
	log.Printf("Unknown error in method %s: %v", method, err)
	return status.Error(codes.Internal, "internal server error")
}

// inventory/internal/model/errors.go (обновленная версия)
package model

import (
	"errors"
	sharedErrors "github.com/green-vlad/mcb/shared/pkg/errors"
)

var (
	ErrPartNotFound = sharedErrors.NewNotFoundError(errors.New("part not found"))
	ErrInvalidUUID  = sharedErrors.NewInvalidArgumentError(errors.New("invalid uuid"))
)

// inventory/internal/repository/inventory/get_part.go (обновленная версия)
package inventory

import (
	"context"

	"github.com/google/uuid"

	"github.com/green-vlad/mcb/inventory/internal/model"
	repoConverter "github.com/green-vlad/mcb/inventory/internal/repository/converter"
)

func (r *repository) GetPart(_ context.Context, partId string) (model.Part, error) {
	id, err := uuid.Parse(partId)
	if err != nil {
		return model.Part{}, model.ErrInvalidUUID
	}
	part, ok := r.parts[id]
	if !ok {
		return model.Part{}, model.ErrPartNotFound
	}
	return repoConverter.PartToModel(part), nil
}

// inventory/internal/api/v1/get_part.go (обновленная версия)
package v1

import (
	"context"

	"github.com/green-vlad/mcb/inventory/internal/converter"
	inventoryv1 "github.com/green-vlad/mcb/shared/pkg/proto/inventory/v1"
)

func (a *api) GetPart(ctx context.Context, req *inventoryv1.GetPartRequest) (*inventoryv1.GetPartResponse, error) {
	part, err := a.inventoryService.GetPart(ctx, req.Uuid)
	if err != nil {
		// Просто возвращаем ошибку, interceptor обработает её автоматически
		return nil, err
	}
	return &inventoryv1.GetPartResponse{
		Part: converter.PartToProto(&part),
	}, nil
}
```

mistandok commented on inventory/internal/api/inventory/v1/get_test.go:
```diff
@@ -0,0 +1,149 @@
+		UpdatedAt:     time.Now(),
+	}
+
+	tests := []struct {
```
suggestion: если один тест - не надо табличный тест реализовывать

mistandok commented on inventory/internal/service/part/get.go:
```diff
@@ -0,0 +1,17 @@
+)
+
+func (s *service) Get(ctx context.Context, partID uuid.UUID) (serviceModel.Part, error) {
+	part, err := s.inventoryRepo.Get(ctx, partID)
```
suggestion: `return s.inventoryRepo.Get(ctx, partID)`

mistandok commented on inventory/internal/repository/model/part.go:
```diff
@@ -0,0 +1,43 @@
+package model
```
thought: не вижу смысла для моделей в in-memory хранилищах на основе мапы. Из плюсов - точно никак на datarace не нарвемся, когда из хранилища возвращаем данные, так как будем при конвертации новые объекты создавать. Если напрямую указатели на модели из сервисного слоя хранить то такой риск остается, надо аккуратнее выстраивать код. 

Тут не ошибка, с точки зрения слоистой архитекруты это верно, когда будем переходить на postgres это также поможет. 

А так это больше мысли вслух на порассуждать.

mistandok commented on order/internal/model/enums.go:
```diff
@@ -0,0 +1,27 @@
+type PaymentMethod string
+
+const (
+	PaymentMethodUNKNOWN       PaymentMethod = "UNKNOWN"
```
thought: кажется что UNKNOWN на уровне сервиса - это уже ошибка, так как если мы не знаем, как клиент заплатил, то не знаем, как ему вернуть деньги или не знаем, как в принципе проводить этот платеж. Можно надо этим подумать.

mistandok commented on order/internal/repository/order/update.go:
```diff
@@ -0,0 +1,25 @@
+	order := r.data[data.UUID]
+	defer r.mu.Unlock()
+
+	switch kind {
```
thought: мне не очень нравится это решение, потому что у заказа может быть 1000 и 1 причина для обновления и обновляться может 1000 и 1 комбинация полей. Я бы просто сделать patch операцию и посылал на обновление целиком сущность заказа, которая бы подменяла текущую. 

Не ставлю как ошибку, так как будем прееходить на реальное хранилище.

mistandok commented on order/internal/service/order/cancel.go:
```diff
@@ -0,0 +1,35 @@
+		OrderStatus: model.OrderStatusCANCELLED,
+	}
+
+	err = s.orderRepo.Update(ctx, updatedInfo, model.OrderUpdateCANCEL)
```
thought: То, что дальше опишу - за рамками курса. Это не надо фиксить и упарываться, у нас упор на микросервисную составляющую, как организовать взаимодействие, поднимать сервисы и т.п, а не про бизнес логику. Так вот:

Ты получил из репозитория заказ, в текущий момент пусть он будет Cancelable. Дальше ты будешь менять его статус и попробуешь сохранить еще раз в бд. Какие могут быть проблемы? проблемы могут быть следующие: в момент, когда ты вычитал заказ из репозитория, его вычитал кто-то еще в параллельном запросе и изменил ему статус быстрее на какой-то другой. Но ты в своем коде про это не узнаешь и изменишь его статус на Cancel. И приведешь базу в неконсистентное состояние.
Как это можно решать? через оптимистичные и пессиместичные блокировки.

Оптимистичные: в заказе появляется версия, ты вычитал заказ, он у тебя имеет версию 1, если ты сохраняешь заказ и у него до сих пор версия 1 - значит его никто не трогал.
Пессимистичные: явная блокировка для изменений. В постгре, к примеру, это select for update, pg_adwisory_lock и т.п..

Эти вещи надо иметь ввиду когда пишешь подобный код

mistandok reviewed (approved):
Привет! в целом слоистая архитектура реализована корректно. Дал несколько рекомендаций, критических ошибок не увидел.

crafty-ezhik committed:
fix: Исправления + рефакторинг

1. Изменено тело функции get в Inventory и Order service
2. Добавлена проверка на метод платежа
